Make cors middleware compliant with the intended spec

alexandrebodin opened the pull request:
<!--
Hello 👋 Thank you for submitting a pull request.

To help us merge your PR, make sure to follow the instructions below:

- Create or update the tests
- Create or update the documentation at https://github.com/strapi/documentation
- Refer to the issue you are closing in the PR description: Fix #issue
- Specify if the PR is ready to be merged or work in progress (by opening a draft PR)

Please ensure you read the Contributing Guide: https://github.com/strapi/strapi/blob/main/CONTRIBUTING.md
-->

### What does it do?

Make cors middleware compliant with the intended spec

derrickmehaffy commented:
building exp to test real world usage: https://github.com/strapi/strapi/actions/runs/8575893524

derrickmehaffy commented:
exp version: `0.0.0-experimental.00cd76085e4bb6c15688e9671f74f517f25926c1`

derrickmehaffy commented:
Running a new one as I'm getting some odd sharp errors, develop might be broken or something :thinking: 

https://github.com/strapi/strapi/actions/runs/8576033020/job/23506249541

alexandrebodin commented on packages/core/strapi/package.json:
```diff
@@ -111,7 +111,7 @@
   },
   "dependencies": {
-    "@koa/cors": "3.4.3",
+    "@koa/cors": "5.0.0",
```
they migrated to ts and drop some node version supports like we did in the past. 

Main change is the default origin which meant having to reconsider the way we did things on our side.

alexandrebodin commented on packages/core/strapi/src/middlewares/cors.ts:
```diff
@@ -46,17 +50,11 @@ export const cors: Common.MiddlewareFactory<Config> = (config) => {
-      const requestOrigin = ctx.headers.origin ?? '';
-      if (whitelist.includes('*')) {
-        return credentials ? requestOrigin : '*';
+      if (Array.isArray(originList)) {
```
explained in the general answer below

derrickmehaffy commented:
New exp version: `0.0.0-experimental.d5b46d578a5c055b8dcc66939e1b5d540976fafb`

alexandrebodin commented:
> @alexandrebodin I'm getting this on the experimental, idk if it has something to do with your changes but something is broken: ![image](https://private-user-images.githubusercontent.com/8593673/320545741-a54baeb9-3118-476f-8a80-cde61fe9228b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI1OTc2OTksIm5iZiI6MTcxMjU5NzM5OSwicGF0aCI6Ii84NTkzNjczLzMyMDU0NTc0MS1hNTRiYWViOS0zMTE4LTQ3NmYtOGE4MC1jZGU2MWZlOTIyOGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MDhUMTcyOTU5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2FhYjAzYjhlZWYyMGI4MDExZjYyN2RhY2RmY2E2ZTIzMjBkNzcxMGJiMGU0NzdjODQzZGM1NDQ3ZDkzMjc2NiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.kNopSd7rJ5CbB8RhVhmqAL2mFgATqGn7jvRKjGjyHDw)

nope not relateds, looks like a sharp binary issue

derrickmehaffy commented:
updating branch and releasing a new exp to fix the sharp issue

derrickmehaffy commented:
New exp with sharp fix (again): https://github.com/strapi/strapi/actions/runs/8634304417/job/23669419178

derrickmehaffy commented:
@alexandrebodin as we discussed in Discord, we are setting the `Access-Control-Allow-Origin` to multiple values which isn't possible so best solution (I think) is to add the following:

- If our origin is an array, check the incoming `origin` header to see if it's in the array, if so set `Access-Control-Allow-Origin` to that origin
- If it's not in the array either we can just set `Access-Control-Allow-Origin` to the first value in the array or some non-sense domain (the only reason I would say the non-sense is we could specify the error message as the domain, not sure if there is a spec for that or not).

derrickmehaffy commented:
New exp: https://github.com/strapi/strapi/actions/runs/8696106150

derrickmehaffy reviewed (approved):
Tested pretty every option related to the middleware origin config and it all looks good :+1:

derrickmehaffy commented:
final exp: `0.0.0-experimental.469f8c0fe2f5ae4a405ec89049706cd7debab0f7`

Pablo-Aldana commented:
This changes seems to break previous cors configuration and I couldn't find documentation about how to upgrade to the new changes.

Right now I have my CORS like this:

```
{
    name: 'strapi::cors',
    config: {
      //enabled: true,
      headers: '*',
      origin: env('CORS'),
    },
  },
  ```

And in the env(CORS) like this but with PROD values:
CORS=http://localhost:1338, http://localhost:8000, http://192.168.68.114:8000, http://192.168.68.114:3000, http://localhost:3000, http://192.168.68.114:1338

As soon as I update to this the strapi version 4.24.0 (https://github.com/strapi/strapi/releases/tag/v4.24.0)or higher, the web stop working due to cors issues, even if the URL is in the cors.

![Screenshot_2024-06-13 23 40 57_studUG](https://github.com/strapi/strapi/assets/9969839/6b1b817f-fa54-45e5-a4ec-66d7037bdb82)


Can anyone point me to the documentation or  update guide for this change?
I can create a new issue if needed

Many thanks

alexandrebodin commented:
With the cors fix we had to do for security we removed the string list support as we don't know if it makes sense to process it this way or not. to fix this for your usecase you should split the string yourself 

`env('CORS').split(', ')`
